doc/neps/nep-0052-python-api-cleanup.rst

```diff
@@ -0,0 +1,331 @@
+   >>> modules
+   ['char', 'compat', 'ctypeslib', 'emath', 'fft', 'lib', 'linalg', 'ma', 'math', 'polynomial', 'random', 'rec', 'testing', 'version']
+   >>> len(modules)
+   14
+
+The above doesn't even include items that are public but have been
+been hidden from ``__dir__``.
+A particularly problematic example of that is ``np.core``,
+which is technically private but heavily used in practice.
+For a full overview of what's consider public, private or a bit in between, see
+`<https://github.com/numpy/numpy/blob/main/numpy/tests/test_public_api.py>`__.
+
+The size of the API and the lacking definition of its boundaries
+incur significant costs:
+
+- **Users find it hard to disambiguate between similarly named
+  functions.**
+
+  Looking for functions with
+  tab completion in IPython, a notebook, or an IDE is a challenge. E.g., type
+  ``np.<TAB>`` and look at the first six items offered: two ufuncs
+  (``abs``, ``add``), one alias (``absolute``), and three functions that are
+  not intended for end-users (``add_docstring``, ``add_newdoc``, ``add_newdoc_ufunc``).
+  As a result, the learning curve for NumPy is steeper than it has to be.
+
+- **Libraries that mimic the NumPy API face significant implementation barriers.**
+
+  For maintainers of NumPy API-compatible array libraries (Dask, CuPy, JAX, PyTorch,
+  TensorFlow, cuNumeric, etc.) and compilers/transpilers (Numba, Pythran,
+  Cython, etc.) there is an implementation cost to each object in the
+  namespace. In practice, no other library has full support for the entire
+  NumPy API, partly because it is so hard to know what to include when faced
+  with a slew of aliases and legacy objects.
+
+- **Teaching NumPy is more complicated than it needs to be.**
+
+  Similarly, a larger API is confusing to learners, who not only have
+  to *find* functions but have to choose *which* functions to use.
+
+- **Developers are hesitant to grow the API surface.**
+
+  This happens even when the changes are warranted, because they are
+  aware of the above concerns.
+
+.. R: TODO: find and link discussion about restructuring namespaces! (e.g.,
+   find the thread with the GUI explorer person)
+
+.. S: Aaron's post re: array API and NumPy 2.0:
+   https://mail.python.org/archives/list/numpy-discussion@python.org/thread/TTZEUKXUICDHGTCX5EMR6DQTYOSDGRV7/#YKBWQ2AP76WYWAP6GFRYMPHZCKTC43KM
+
+The scope of this NEP includes:
+
+- Deprecating or removing functionality that is too niche for NumPy, not
+  well-designed, superceded by better alternatives, or otherwise a candidate
+  for removal.
+- Clearly separating public from private NumPy API by use of underscores.
+- Restructuring the NumPy namespaces to be easier to understand and navigate.
+
+Out of scope for this NEP are:
+
+- Introducing new functionality or performance enhancements.
+
+TBD whether in or out of scope for this NEP:
+
+- Whether or not to clean up some of the more niche ``ndarray`` methods (this
```

Comment:
In favour.

Reply:
Included this now, with five example methods to clean up..